torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -0,0 +1,613 @@
+        self.needs_pre_backward_unshard = False
+
+    def _prefetch_unshard(self):
+        if self._training_state in (
+            TrainingState.PRE_BACKWARD,
+            TrainingState.POST_BACKWARD,
+        ):
+            if self._post_forward_index is None:
+                return
+            target_index = self._post_forward_index - 1
+            if target_index < 0:
+                return
+            target_fsdp_param_group = self.post_forward_order[target_index]()
+            assert target_fsdp_param_group is not None, "Weakref deallocated"
+            if (
+                target_fsdp_param_group.needs_pre_backward_unshard
+                and not target_fsdp_param_group._prefetched
+            ):
+                with torch.profiler.record_function("FSDP::backward_prefetch"):
+                    with target_fsdp_param_group.use_training_state(
+                        TrainingState.PRE_BACKWARD
+                    ):
+                        target_fsdp_param_group._unshard()
+                target_fsdp_param_group._prefetched = True
+
+    # State Dict #
+    def _pre_save_state_dict_hook(
+        self, module: nn.Module, *unused_args, **unused_kwargs
+    ):
+        if module not in self._module_to_fsdp_param_refs:
+            raise FSDPInternalError(
+                f"Module not found in module to FSDP parameter mapping: {module}"
+            )
+        module_fsdp_params = self._module_to_fsdp_param_refs[module]
+        for fsdp_param_ref in module_fsdp_params:
+            fsdp_param = fsdp_param_ref()
+            assert fsdp_param is not None, "FSDPParam deallocated"
+            fsdp_param.to_sharded()
+
+    def _pre_load_state_dict_hook(
+        self, module: nn.Module, *unused_args, **unused_kwargs
+    ):
+        if module not in self._module_to_fsdp_param_refs:
+            raise FSDPInternalError(
+                f"Module not found in module to FSDP parameter mapping: {module}"
+            )
+        module_fsdp_params = self._module_to_fsdp_param_refs[module]
+        for fsdp_param_ref in module_fsdp_params:
+            fsdp_param = fsdp_param_ref()
+            assert fsdp_param is not None, "FSDPParam deallocated"
+            fsdp_param.to_sharded()
+
+    # Utilities #
+    @contextlib.contextmanager
+    def use_training_state(self, training_state: TrainingState):
+        old_training_state = self._training_state
+        self._training_state = training_state
+        try:
+            yield
+        finally:
+            self._training_state = old_training_state
+
+    def set_reduce_scatter_grads(self, reduce_scatter_grads: bool):
+        self._reduce_scatter_grads = reduce_scatter_grads
+
+    def set_all_reduce_grads(self, all_reduce_grads: bool):
+        self._all_reduce_grads = all_reduce_grads
+
+    def set_wait_for_grad_sync(self, wait_for_grad_sync: bool):
+        self._wait_for_grad_sync = wait_for_grad_sync
+
+    # Hook Registration #
+    def _register_post_backward_hook(
```

Comment:
This is the convoluted logic to extract the input tensors that require gradient from the module's forward inputs and to pass them through a custom autograd function that is an identity in forward and runs the FSDP-defined post-backward hook in backward.

I am not sure if there is a better way to do this. I do not want to install any global state, so I was hesitant to use full module backward hooks.